### PR TITLE
qt: Avoid potential -Wdeprecated-enum-enum-conversion warnings

### PR DESCRIPTION
**Pull request by hebasto**
This PR is related to bitcoin/bitcoin#24169. It adjusts code in order to avoid `-Wdeprecated-enum-enum-conversion` warnings instead of disabling them.

Could be tested with gcc 11.2.

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -449,6 +449,7 @@ if test "$CXXFLAGS_overridden" = "no"; then
                         [AC_LANG_SOURCE([[struct A { virtual void f(); }; struct B : A { void f() final; };]])])
   AX_CHECK_COMPILE_FLAG([-Wunreachable-code-loop-increment], [WARN_CXXFLAGS="$WARN_CXXFLAGS -Wunreachable-code-loop-increment"], [], [$CXXFLAG_WERROR])
   AX_CHECK_COMPILE_FLAG([-Wimplicit-fallthrough], [WARN_CXXFLAGS="$WARN_CXXFLAGS -Wimplicit-fallthrough"], [], [$CXXFLAG_WERROR])
+  AX_CHECK_COMPILE_FLAG([-Wdeprecated-enum-enum-conversion], [WARN_CXXFLAGS="$WARN_CXXFLAGS -Wdeprecated-enum-enum-conversion"], [], [$CXXFLAG_WERROR])
```
Comment (fanquake) [inline configure.ac:452]: Looks like his is enabled by default when targeting C++20 in GCC, and by default in Clang, so I don't think there's a need to add this explicitly.

Comment (MarcoFalke) [reply configure.ac:452]: I presume it is not enabled by default in C++17, which is used to compile Bitcoin Core (and will be for the coming months/years)?

File: src/qt/bitcoingui.cpp [hunk 2]
```diff
@@ -251,28 +252,28 @@ void BitcoinGUI::createActions()
     overviewAction->setStatusTip(tr("Show general overview of wallet"));
     overviewAction->setToolTip(overviewAction->statusTip());
     overviewAction->setCheckable(true);
-    overviewAction->setShortcut(QKeySequence(Qt::ALT + Qt::Key_1));
+    overviewAction->setShortcut(QKeySequence(tr("Alt+1")));
```
Comment (MarcoFalke) [inline src/qt/bitcoingui.cpp:255]: Isn't `tr` used to mark messages for translation? If yes, why is this needed here? If this is needed to convert to `QString`, why not call `QString{}` instead?

File: configure.ac (hunk 1 above)
Comment (fanquake) [reply configure.ac:452]: It's enabled by default in Clang, which should be good enough coverage for us.
In GCC, when targetting pre-C++20, it's enabled by `-Wenum-conversion`, which is part of `-Wextra`, which we already enable. Although in that case it may only be applied to C code.

File: src/qt/bitcoingui.cpp (hunk 2 above)
Comment (hebasto) [reply src/qt/bitcoingui.cpp]: > Isn't `tr` used to mark messages for translation?

Yes, it is.

> If yes, why is this needed here? If this is needed to convert to `QString`, why not call `QString{}` instead?

No need for translation here. Updated in the recent push.

Thanks!

File: configure.ac (hunk 1 above)
Comment (hebasto) [reply configure.ac:452]: > I presume it is not enabled by default in C++17

Correct.

Comment (MarcoFalke) [reply configure.ac:452]: > It's enabled by default in Clang, which should be good enough coverage for us.

For me it is only enabled with c++20. Do you have steps to reproduce?

Comment (fanquake) [reply configure.ac:452]: > Do you have steps to reproduce?

No. I assumed Clang might be enabling with any standard, given they [don't specify otherwise](https://clang.llvm.org/docs/DiagnosticsReference.html#wdeprecated-enum-enum-conversion), like GCC does. However, I'm still a NACK on adding this. It's already enabled by default when it's needed, it's only available on fairly "new" compilers, so having everyone building with older ones check for availability is pointless, and I'd prefer that our configure checks don't return to being a grab-bag of checks for random flags. In any case, given the only existence of warning producing code, are a few instances in the GUI, I'd say the likelihood of any warning producing code being introduced is a also very low.

Comment (hebasto) [reply configure.ac:452]: > I'm still a NACK on adding this.

An additional minor pro for adding this line is simplicity of testing this PR by reviewers using gcc.

File: src/qt/bitcoingui.cpp [hunk 3]
```diff
@@ -251,28 +252,28 @@ void BitcoinGUI::createActions()
     overviewAction->setStatusTip(tr("Show general overview of wallet"));
     overviewAction->setToolTip(overviewAction->statusTip());
     overviewAction->setCheckable(true);
-    overviewAction->setShortcut(QKeySequence(Qt::ALT + Qt::Key_1));
+    overviewAction->setShortcut(QKeySequence(QString{"Alt+1"}));
```
Comment (promag) [inline src/qt/bitcoingui.cpp:255]: 0fea3846d0d6ec69dadd8bf043afd3c25971d07f

nit, here and below, use `QStringLiteral`, see https://doc.qt.io/qt-5/qstring.html#more-efficient-string-construction

File: src/qt/guiutil.cpp [hunk 4]
```diff
@@ -414,7 +415,7 @@ void bringToFront(QWidget* w)
 
 void handleCloseWindowShortcut(QWidget* w)
 {
-    QObject::connect(new QShortcut(QKeySequence(Qt::CTRL + Qt::Key_W), w), &QShortcut::activated, w, &QWidget::close);
+    QObject::connect(new QShortcut(QKeySequence(QShortcut::tr("Ctrl+W")), w), &QShortcut::activated, w, &QWidget::close);
```
Comment (promag) [inline src/qt/guiutil.cpp:418]: 0fea3846d0d6ec69dadd8bf043afd3c25971d07f

`QObject::tr()`

**Review by promag — COMMENTED** [review]
Concept ACK Qt change.

**Comment by hebasto** [comment]
Updated 38cffbb301e8d2ddf8dbc12afee99e76d8b54ebd -> 103b8dd8b89c772b7f6d097383e83d742b5ed55c ([pr24624.03](https://github.com/hebasto/bitcoin/commits/pr24624.03) -> [pr24624.04](https://github.com/hebasto/bitcoin/commits/pr24624.04), [diff](https://github.com/hebasto/bitcoin/compare/pr24624.03..pr24624.04)):

- addressed @promag's comments

File: src/qt/bitcoingui.cpp (hunk 3 above)
Comment (hebasto) [reply src/qt/bitcoingui.cpp]: Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/24624#issuecomment-1074850209).

File: src/qt/guiutil.cpp (hunk 4 above)
Comment (hebasto) [reply src/qt/guiutil.cpp]: The initial idea was to put this translation into the context named "QShortcut", rather "QObject". But context name does not go from `src/qt/locale/bitcoin_en.ts` into the Transifex environment. So [accepting](https://github.com/bitcoin/bitcoin/pull/24624#issuecomment-1074850209) your suggestion.

**Comment by MarcoFalke** [comment]
Can you explain why any of this needs translation? Is it displayed anywhere to the user?

**Comment by hebasto** [comment]
> Can you explain why any of this needs translation? Is it displayed anywhere to the user?

Screenshot from 2022-03-22 09-16-01

File: configure.ac (hunk 1 above)
Comment (fanquake) [reply configure.ac:452]: > An additional minor pro for adding this line is simplicity of testing this PR by reviewers using gcc.

I don't really think we should be adding warning flags to our build system for the sake of making testing PRs easier. If anyone wants to test this change, they can do `./configure CXXFLAGS="-Wdeprecated-enum-enum-conversion"` using a compiler that supports the option.

**Comment by hebasto** [comment]
Updated 103b8dd8b89c772b7f6d097383e83d742b5ed55c -> acd98adaf1d83b71eda235c49d41a92f30c16313 ([pr24624.04](https://github.com/hebasto/bitcoin/commits/pr24624.04) -> [pr24624.05](https://github.com/hebasto/bitcoin/commits/pr24624.05), [diff](https://github.com/hebasto/bitcoin/compare/pr24624.04..pr24624.05)):

- addressed @fanquake's [comment](https://github.com/bitcoin/bitcoin/pull/24624#discussion_r832060992)

File: configure.ac (hunk 1 above)
Comment (hebasto) [reply configure.ac]: [Updated](https://github.com/bitcoin/bitcoin/pull/24624#issuecomment-1075072189).

**Review by fanquake — APPROVED** [review]
untested ACK acd98adaf1d83b71eda235c49d41a92f30c16313 - thanks.

**Comment by MarcoFalke** [comment]
Approach ACK acd98adaf1d83b71eda235c49d41a92f30c16313

For testing locally, #24169 will need to be merged.

**Comment by fanquake** [comment]
> For testing locally, https://github.com/bitcoin/bitcoin/pull/24169 will need to be merged.

Why? You can just do `./configure CXXFLAGS="-Wdeprecated-enum-enum-conversion"` with a supporting compiler.

**Review by promag — COMMENTED** [review]
Code review ACK acd98adaf1d83b71eda235c49d41a92f30c16313.
